### PR TITLE
Add Ability to Run as Full Simulator

### PR DESCRIPTION
**Pull request by carlaKC**
This PR is a WIP that adds the ability to run SimLN without any real lightning nodes, instead providing "simulated nodes" that copy the the routing policies and liquidity constraints of real nodes. The last commit hard-codes a graph to run this "full simulation" on as an example, whereas in reality we'd load from a json description of the desired topology/policies.

### High Level Overview
Lightning implementations in SimLN are abstracted using the [LightningNode](https://github.com/bitcoin-dev-project/sim-ln/blob/main/sim-lib/src/lib.rs#L180) trait. The steps that we follow to support a simulated implementation of this trait (SimNode) are as follows: 
- Run the simulator with a `–from_graph option`, which allows starting the simulator with a graph file (format tbd) that provides channel policies and topology (not included in this PR). 
- Use a single, top-level `SimGraph` struct to manage: 
  - A hashmap of simulated channels available
  - Propagation of payments through the simulated channels
- Implement [LightningNode](https://github.com/bitcoin-dev-project/sim-ln/blob/main/sim-lib/src/lib.rs#L180) on a SimNode struct which: 
  - Handles implementation of send/tracking functionality
  - Points to the coordinating `SimGraph` struct to manage payment dispatch.

### Graph Abstraction
Management of payments through the simulated network is separated managed with the following layers:

- SimGraph: high level coordinator that keeps a hashmap of each simulated channel (see below) in the network, along with their forwarding policies. Responsible for coordinating addition/removal of HTLCs as they propagate through the network.
- SimNode: implementation of the LightningNode trait which rely on the SimGraph to propagate payments on their behalf. 

<p align="center">
Screenshot 2024-01-30 at 4 43 35 PM
</p>

These structs are primarily used to provide send/track payment APIs, receiving updates on the state of their payments from SimGraph.

### Channel State Machine

Implementation of channel state is broken down into three layers:
- Channel Policy: representation of the directional policy of a participant (fees etc)
- Channel State: tracks the “live” state of the outgoing state of the channel
- Simulated Channel: tracks state for each direction of the channel, and high level information

<p align="center">
Screenshot 2024-01-30 at 4 19 13 PM
</p>

Each channel state is responsible for handling tracking of liquidity and HTLC limits in one direction, and the simulated channel is responsible for enforcing sanity checks across the channel (eg, that we don’t exceed our capacity).

An example state machine update is provided in the table below:
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-0a21a89f-7fff-4cd6-26e7-27365e6cec71"><div dir="ltr" style="margin-left:0pt;" align="left">
Step | Alice Channel State | Bob Channel State
-- | -- | --
Channel Opened by Alice | Local_balance: 100_000 / In_Flight: nil | Local_balance: 0 / In_Flight: Nil
Alice sends 100 sats to Bob | Local_balance: 999_000 / In_Flight: 100 sats | Local_balance: 0 / In_Flight: Nil
Alice payment succeeds | Local_balance: 999_000 / In_Flight: nil | Local_balance: 100 / In_Flight: Nil
Bob sends 50 sats to Alice | Local_balance: 999_000 / In_Flight: nil | Local_balance: 50 / In_Flight: 50
Bob payment fails | Local_balance: 999_000 / In_Flight: nil | Local_balance: 100 / In_Flight: nil

</div></b>

### tl;dr
<p align="center">
Screenshot 2024-01-30 at 4 01 34 PM
</p>

### CODE REVIEW

File: sim-lib/Cargo.toml [hunk 1]
```diff
@@ -15,7 +15,9 @@ expanduser = "1.2.2"
 serde = { version="1.0.183", features=["derive"] }
 serde_json = "1.0.104"
 bitcoin = { version = "0.30.1", features=["serde"] }
-lightning = { version = "0.0.116" }
+# LDK is on a different version to bitcoin to us, which leads to some issues, import so we can use its types.
```
Comment (enigbe) [inline sim-lib/Cargo.toml:18]: Are there reasons why `bitcoin_ldk` types are being used as opposed to sticking with types from `bitcoin`?

File: sim-lib/src/sim_node.rs [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+                | ForwardingError::ChannelNotFound(_)
+                | ForwardingError::NodeNotFound(_)
+                | ForwardingError::PaymentHashExists(_)
+                | ForwardingError::PaymentHashNotFound(_)
+                | ForwardingError::SanityCheckFailed(_, _, _)
+        )
+    }
+}
+
+impl Display for ForwardingError {
```
Comment (enigbe) [inline sim-lib/src/sim_node.rs:92]: Why not derive `thiserror::Error`? It would reduce LOC

File: sim-lib/src/sim_node.rs [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+        SimulatedChannel {
+            capacity_msat,
+            short_channel_id,
+            node_1: ChannelState::new(node_1, capacity_msat),
+            node_2: ChannelState::new(node_2, capacity_msat),
+        }
+    }
+}
+
+impl SimulatedChannel {
```
Comment (enigbe) [inline sim-lib/src/sim_node.rs:281]: Why have another implementation of `SimulatedChannel` if there isn't a huge difference between the concerns in both `impl`s?

File: sim-lib/src/sim_node.rs [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+                self.policy.fee_rate_prop,
+                expected_fee,
+            ));
+        }
+
+        Ok(())
+    }
+
+    fn check_policy(&self, htlc: &Htlc) -> Result<(), ForwardingError> {
+        if htlc.amount_msat > self.local_balance_msat {
```
Comment (enigbe) [inline sim-lib/src/sim_node.rs:182]: Could cut on LOC with a `macro!(conditional_expr, error_val)`

**Review by enigbe — COMMENTED** [review]
Added a handful of small comments on code style and a question about using types from LDK's bitcoin library. 

This would benefit from documentation comments across several structs and methods but expect they'd be updated before this is ready

**Comment by okjodom** [comment]
oh, my! 🎆 🥇

File: sim-lib/src/sim_node.rs (hunk 3 above)
Comment (carlaKC) [reply sim-lib/src/sim_node.rs:281]: Mistake!

File: sim-lib/Cargo.toml (hunk 1 above)
Comment (carlaKC) [reply sim-lib/Cargo.toml:18]: Long and painful story that's now resolved, but for prosperity: 

We were relying of a different version of the `bitcoin` dependency to LDK:
* `sim-ln` -> `bitcoin-30`
* `sim-ln` -> `lightning-117` -> `bitocin-29`

We couldn't downgrade `bitcoin-30`, because we're using a feature it it. But to use `lightning-117` (specifically in pathfinding) we needed to provide the public key struct type in `bitcoin-29`. So this alias allowed me to import both types. This was messy and gross and thankfully solved by another LDK release that updated the bitcoin version.

File: sim-lib/src/sim_node.rs (hunk 2 above)
Comment (carlaKC) [reply sim-lib/src/sim_node.rs:92]: Because I forgot that it existed, will update!

**Comment by carlaKC** [comment]
Pushed some additional docs + cleanup, no semantic changes!

**Comment by carlaKC** [comment]
Pushed a [off by one](https://github.com/bitcoin-dev-project/sim-ln/compare/4dc545526a53a7c231b4fb03ebdaf043fd5d3a8a..3fb2d0d550ada385e023df65bcdb8f3999d9af17) bug that I caught writing tests!

**Review by sr-gi — COMMENTED** [review]
I think the design makes sense overall. I may need to give this a second pass down the line (the diff is huge 😅 ) but my current understanding is as follows:

`SimNode` is the main class that mimics our current node implementations, and hence it implements `LightningNode`. It has a shared reference to a network (`Arc<Mutex<SimGraph>>`) that represents the simulated network. The network includes all channels, which have a breakdown in classes all the way down to the channel policies. The construction of a simulation works as usual: we create a list of clients that are instances that implement `LightningNode` (in this case `SimNode`s), when a payment is made, the simulated node accesses the graph via its shared pointer and triggers all the state transitions, which are managed by the `SimGraph`.

I left a bunch of suggestions, most (if not all of them) can be found here: https://github.com/sr-gi/sim-ln/commit/67f810e7206595e3b1102e1f53a890732da673e3. Feel free to cherry-pick as desired.

Some general comments:

- I don't think 604019da218bf50acb51e280adabd68452c6133e should be needed. We can have a different constructor for the simulator when calling it with a simulated graph (`Simulator::with_sim_channels` or something) that receives all the usual but instead of a map of clients, receives a collection of simulated channels. That constructor could have the functionality of `ln_node_from_graph` to build the "clients", plus construct (https://github.com/sr-gi/sim-ln/commit/67f810e7206595e3b1102e1f53a890732da673e3#diff-f891364aff98268ac6d94109bd61594c23af50e37e547fec9b0ea44a816459a0R390-R421). Doing this gets rids of a bunch of imports, plus having to define the triggers in the CLI and so on.

- I think d2ea9363342d3d828f4858dc26fcceb2c57d8f3b should be squashed, I don't think it's meaningful enough to be a commit on its own (but up to you in the end)

**Comment by carlaKC** [comment]
Thanks for the mega-review @sr-gi 🫶 

Not sure if this is helpful, but I've pushed a single fixup for each commit with the suggested changes (each fixup directly follows its original commit) with the hope of making round two a little easier! If it's not useful I'll just go ahead and squash. 

Added a few responses inline where the big-picture stuff needs some discussion.

**Comment by carlaKC** [comment]
Also, I've started on some unit test coverage for this but going to put it in as a separate PR because this one is already far too big :')

File: sim-lib/src/sim_node.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    /// Zero amount htlcs are in valid in the protocol.
+    ZeroAmountHtlc,
+    /// The outgoing channel id was not found in the network graph.
+    ChannelNotFound(u64),
+    /// The node pubkey provided was not associated with the channel in the network graph.
+    NodeNotFound(PublicKey),
+    /// The channel has already forwarded a HTLC with the payment hash provided (to be removed when MPP support is
+    /// added).
+    PaymentHashExists(PaymentHash),
+    /// A htlc with the payment hash provided could not be found to resolve.
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: I had to think about this one twice before nitpicking (and double-check it with @valentinewallace :P)

File: sim-lib/src/sim_node.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    ) -> Result<(), ForwardingError> {
+        if cltv_delta < self.policy.cltv_expiry_delta {
+            return Err(ForwardingError::InsufficientCltvDelta(
+                cltv_delta,
+                self.policy.cltv_expiry_delta,
+            ));
+        }
+
+        // As u64 will round expected fee down to nearest msat (this is what the protocol dictates).
+        let expected_fee = (self.policy.base_fee as f64
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: Nice!

File: sim-lib/src/sim_node.rs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    fn remove_outgoing_htlc(
+        &mut self,
+        hash: PaymentHash,
+        success: bool,
+    ) -> Result<Htlc, ForwardingError> {
+        match self.in_flight.remove(&hash) {
+            Some(v) => {
+                // If the HTLC failed, pending balance returns to local balance.
+                if !success {
+                    self.local_balance_msat += v.amount_msat
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: The inner scope returns the result of the `+=`, which is the unit type anyway, so not capturing it has no real effect. That's why I suggested it as a nit.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d7fb5c195dd73d04fcc1f0d900504550

File: sim-lib/src/sim_node.rs [hunk 8]
```diff
@@ -202,3 +202,142 @@ impl ChannelState {
...
+    node_1: ChannelState,
+    node_2: ChannelState,
+}
+
+impl SimulatedChannel {
+    /// Creates a new channel with the capacity and policies provided. The total capacity of the channel is evenly split
+    /// between the channel participants (this is an arbitrary decision).
+    pub fn new(
+        capacity_msat: u64,
+        short_channel_id: u64,
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs:277]: Even if we just add it as an alias without re-defining any method it would make things more readable, especially for the error enums where multiple params are passed. `ChannelNotFound(scid)` reads better than `ChannelNotFound(u64)` for instance.

Regarding proper formatting, I meant to do it like: `A:B:C` or `AxBxC`, but if you think we may not need to manually parse that often from the logs it may not be necessary.

File: sim-lib/src/sim_node.rs [hunk 9]
```diff
@@ -229,23 +229,42 @@ impl SimulatedChannel {
         }
     }
 
-    /// Adds a htlc to the appropriate side of the simulated channel, checking its policy and balance are okay.
-    fn add_htlc(&mut self, node: PublicKey, htlc: Htlc) -> Result<(), ForwardingError> {
-        if htlc.amount_msat == 0 {
-            return Err(ForwardingError::ZeroAmountHtlc);
+    fn get_mut_node(&mut self, pubkey: &PublicKey) -> Result<&mut ChannelState, ForwardingError> {
```
Comment (sr-gi) [inline sim-lib/src/sim_node.rs:289]: nit: the usual naming convention for mutable methods is to have `_mut` at the end

refs:

https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get_mut
https://doc.rust-lang.org/std/borrow/trait.BorrowMut
https://doc.rust-lang.org/std/ops/trait.DerefMut.html
https://doc.rust-lang.org/std/ops/trait.IndexMut.html

File: sim-lib/src/sim_node.rs [hunk 10]
```diff
@@ -202,3 +202,142 @@ impl ChannelState {
...
+        }
+
+        Ok(())
+    }
+
+    /// Removes a htlc from the appropriate size of the simulated channel, settling balances across channel sides
+    /// based on the success of the htlc.
+    fn remove_htlc(
+        &mut self,
+        incoming_node: PublicKey,
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: Sounds good to me

File: sim-lib/src/sim_node.rs [hunk 11]
```diff
@@ -202,3 +202,142 @@ impl ChannelState {
...
+						if $success {
+							$self.$receiver.local_balance_msat += htlc.amount_msat;
+						}
+
+						return $self.sanity_check();
+					},
+					Err(e) => Err(e),
+				}
+			};
+		}
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: That makes sense to me :)

File: sim-lib/src/sim_node.rs [hunk 12]
```diff
@@ -202,3 +202,142 @@ impl ChannelState {
...
+
+        if incoming_node == self.node_1.policy.pubkey {
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply sim-lib/src/sim_node.rs]: I guess we could always add a complementary function if we need to, that returns the opposite `ChannelState` of the node being provided

**Review by sr-gi — COMMENTED** [review]
Looks good! 🎉 

I can confirm most of the comments have been addressed or refuted, just left a few comments on the open discussion plus things that may have been passed unnoticed. 

I'm happy with the fixups to be squashed at this point.

File: sim-lib/src/sim_node.rs [hunk 13]
```diff
@@ -341,3 +360,225 @@ impl SimulatedChannel {
...
+            .dispatch_payment(self.info.pubkey, route, preimage, sender);
+
+        Ok(payment_hash)
+    }
+
+    /// track_payment blocks until a payment outcome is returned for the payment hash provided, or the shutdown listener
+    /// provided is triggered. This call will fail if the hash provided was not obtained by calling send_payment first.
+    async fn track_payment(
+        &mut self,
+        hash: PaymentHash,
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply sim-lib/src/sim_node.rs:563]: To remind us! https://github.com/bitcoin-dev-project/sim-ln/issues/169

File: sim-lib/src/sim_node.rs [hunk 14]
```diff
@@ -596,6 +603,331 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
...
+    let mut outgoing_amount = route.fee_msat() + route.final_value_msat();
+    let mut outgoing_cltv = route
+        .hops
+        .iter()
+        .fold(0, |sum, value| sum + value.cltv_expiry_delta);
+
+    let mut fail_idx = None;
+
+    for (i, hop) in route.hops.iter().enumerate() {
+        let mut node_lock = nodes.lock().await;
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply sim-lib/src/sim_node.rs:833]: Will add a comment here 👍

File: sim-cli/src/main.rs [hunk 15]
```diff
@@ -199,13 +199,15 @@ async fn main() -> anyhow::Result<()> {
...
 
+    let (shutdown_trigger, shutdown_listener) = triggered::trigger();
     let sim = Simulation::new(
         clients,
         validated_activities,
         cli.total_time,
         cli.expected_pmt_amt,
         cli.capacity_multiplier,
         write_results,
+        (shutdown_trigger, shutdown_listener),
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply sim-cli/src/main.rs]: Oh very nice, I think I'm going to leave this for a follow up because I'll drop the commit where we actually use the simulator before merge (to be followed up with surfacing the config options to use it).

**Comment by carlaKC** [comment]
Squashed + addressed last comments. I've removed the commit which actually uses the simulator, because we need to add a configuration option for an end user to provide that format (`--simulate-network`, or something like that). 

Will follow up with a PR adding tests + that option.

**Comment by sr-gi** [comment]
ACK [f199c2f](https://github.com/bitcoin-dev-project/sim-ln/pull/165/commits/f199c2f6fa930efd0f22106db660434899e47f9e)

Last comment, feel free to add it as a follow-up if you find this useful:

You can add a couple of utility functions to `ShortChannelID` to convert back and forth from/to u64, that way you can simply call `x.into()` in the code when converting, without having to even use the wrappers/inner values:

```rust
/// Utility function to easily convert from u64 to `ShortChannelID`
impl From<u64> for ShortChannelID {
    fn from(value: u64) -> Self {
        ShortChannelID(value)
    }
}

/// Utility function to easily convert `ShortChannelID` into u64
impl From<ShortChannelID> for u64 {
    fn from(scid: ShortChannelID) -> Self {
        scid.0
    }
}
```

Here are a couple of examples:

```diff
-short_channel_id: channel.short_channel_id.0,
+short_channel_id: channel.short_channel_id.into(),
```

```diff
-let scid = ShortChannelID(hop.short_channel_id);
+let scid = hop.short_channel_id.into();
```

```diff
-return Err(ForwardingError::ChannelNotFound(ShortChannelID(
                    hop.short_channel_id,
                )))
+return Err(ForwardingError::ChannelNotFound(
                    hop.short_channel_id.into(),
                ))
```

**Review by sr-gi — APPROVED** [review]

**Comment by carlaKC** [comment]
> You can add a couple of utility functions to `ShortChannelID` to convert back and forth from/to u64, that way you can simply call `x.into()` in the code when converting, without having to even use the wrappers/inner values:

Added in [last push](https://github.com/bitcoin-dev-project/sim-ln/compare/f199c2f6fa930efd0f22106db660434899e47f9e..f15590477057eb14cbeae4a38b836214a0f5e6e7)!

**Comment by carlaKC** [comment]
Thanks for the great review on this @sr-gi 🫶🫶🫶

**Comment by sr-gi** [comment]
Anytime :D
